src/wh_server_crypto.c

```diff
@@ -79,6 +78,87 @@ static int hsmLoadKeyRsa(whServerContext* server, RsaKey* key, whKeyId keyId)
     }
     return ret;
 }
+#ifdef WOLFSSL_KEY_GEN
+static int hsmCryptoRsaKeyGen(whServerContext* server, whPacket* packet,
+    uint16_t* size)
+{
+    int ret;
+    whKeyId keyId = WOLFHSM_KEYID_ERASED;
+    /* init the rsa key */
+    ret = wc_InitRsaKey_ex(server->crypto->rsa, NULL, server->crypto->devId);
+    /* make the rsa key with the given params */
+    if (ret == 0) {
+        ret = wc_MakeRsaKey(server->crypto->rsa,
+            (word32)packet->pkRsakgReq.size, (long)packet->pkRsakgReq.e,
+            server->crypto->rng);
+    }
+    /* cache the generated key, data will be blown away */
+    if (ret == 0)
+        ret = hsmCacheKeyRsa(server, server->crypto->rsa, &keyId);
+    wc_FreeRsaKey(server->crypto->rsa);
+    if (ret == 0) {
+        /* set the assigned id */
+        packet->pkRsakgRes.keyId = (keyId & WOLFHSM_KEYID_MASK);
+        *size = WOLFHSM_PACKET_STUB_SIZE + sizeof(packet->pkRsakgRes);
+    }
+    return ret;
+}
+#endif /* WOLFSSL_KEY_GEN */
+
+static int hsmCryptoRsaFunction(whServerContext* server, whPacket* packet,
+    uint16_t* size)
+{
+    int ret;
+    word32 len;
+    /* in and out are after the fixed size fields */
+    byte* in = (uint8_t*)(&packet->pkRsaReq + 1);
+    byte* out = (uint8_t*)(&packet->pkRsaRes + 1);
+    /* init rsa key */
+    ret = wc_InitRsaKey_ex(server->crypto->rsa, NULL, server->crypto->devId);
```

Comment:
Future memory reduction cleanup:  Replace all crypto->rsa and other temporary states that are not needed during inter-request times.  I think the only state that is necessary in the crypto context (which is shared among all of the server contexts) is the Rng.  I really like the usage of the keyslot cache to hold data necessary to maintain state between subsequent calls, rather than preallocating them, even in a union.  No change now.

Reply:
I initially had them all on the stack, not sure if we moved it because we hadn't planned out the multiple server context solution but I will move it back onto the stack in another PR